rhel/vex/fetcher.go

```diff
@@ -0,0 +1,288 @@
+			}
+			bc.WriteByte('\n')
+			cw.Write(bc.Bytes())
+			buf.Reset()
+			bc.Reset()
+			entriesWritten++
+		}
+		if err != io.EOF {
+			return nil, hint, fmt.Errorf("error reading tar contents: %w", err)
+		}
+
+		zlog.Debug(ctx).
+			Str("updater", u.Name()).
+			Int("entries written", entriesWritten).
+			Msg("finished writing compressed data to spool")
+
+	}
+
+	uri, err := u.url.Parse("changes.csv")
+	if err != nil {
+		return nil, hint, err
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, uri.String(), nil)
+	if err != nil {
+		return nil, hint, err
+	}
+	if fp.changesEtag != "" {
+		req.Header.Add("If-None-Match", fp.changesEtag)
+	}
+	res, err := u.client.Do(req)
+	if res != nil {
+		defer res.Body.Close()
+	}
+	if err != nil {
+		return nil, hint, err
+	}
+
+	switch res.StatusCode {
+	case http.StatusOK:
+		if t := fp.changesEtag; t == "" || t != res.Header.Get("etag") {
+			break
+		}
+		fallthrough
+	case http.StatusNotModified:
+		// We could return driver.Unchanged here but we don't know for sure. Return the
+		// file that may have data read from the compressed file in it.
+		return f, hint, nil
+	default:
+		return nil, hint, fmt.Errorf("unexpected response from changes.csv: %s", res.Status)
+	}
+	fp.changesEtag = res.Header.Get("etag")
+
+	rd := csv.NewReader(res.Body)
+	rd.FieldsPerRecord = 2
+	rd.ReuseRecord = true
+	var (
+		l       = 0
+		buf, bc bytes.Buffer
+	)
+	rec, err := rd.Read()
+	for ; err == nil; rec, err = rd.Read() {
+		if len(rec) != 2 {
+			return nil, hint, fmt.Errorf("could not parse changes.csv file")
+		}
+		var err error
+		e := Entry{}
+		e.Path += rec[0] // This += should result in us getting a copy.
+		year, err := strconv.ParseInt(path.Dir(e.Path), 10, 64)
+		if err != nil {
+			return nil, hint, fmt.Errorf("error parsing year %w", err)
+		}
+		if year < lookBackToYear {
+			continue
+		}
+		e.updatedTime, err = time.Parse(time.RFC3339, rec[1])
+		if err != nil {
+			return nil, hint, fmt.Errorf("line %d: %w", l, err)
+		}
+		if e.updatedTime.After(fp.requestTime) {
+			en := e
+			advisoryURI, err := u.url.Parse(en.Path)
+			if err != nil {
+				return nil, hint, err
+			}
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, advisoryURI.String(), nil)
+			if err != nil {
+				return nil, hint, fmt.Errorf("error creating advisory request %w", err)
+			}
+
+			res, err := u.client.Do(req)
+			if err != nil {
+				return nil, hint, fmt.Errorf("error making advisory request %w", err)
+			}
+			if res.StatusCode != http.StatusOK {
+				var b strings.Builder
+				if _, err := io.Copy(&b, res.Body); err != nil {
+					zlog.Warn(ctx).Err(err).Msg("additional error while reading error response")
+				} else {
+					zlog.Warn(ctx).Str("response", b.String()).Msg("received error response")
+				}
+				return nil, hint, fmt.Errorf("unexpected response from advisary URL: %s %s", res.Status, req.URL)
+			}
+
+			_, err = buf.ReadFrom(res.Body)
```

Comment:
so does this cover the case where this is the first time we read the VEX data and there is a newer version of a few files? This will overwrite data for those files?

Reply:
At the moment we'll spool the compressed file to disk then append the updated files, the parser takes care of only persisting the latest version.